File: Server/Node.py
import socketio
import os
import sys
import traceback


class Node:
    def __init__(self, node_name, url="http://localhost:5000", skip_connection=False):
        self.sio = socketio.Client()
    
        self.node_name = node_name
        self.topics = {}
        self.connected = False
        
        # connect to server and register node name
        
        @self.sio.event
        def connect():
            print(f"Node {node_name} connected")
            self.sio.emit("connect_node", node_name)
            self.connected = True

        # receive and handle data using topic subscription callback
        @self.sio.event
        def get_data(data):
            topic, payload = data

            topicFunction = self.topics.get(topic, None)
            if topicFunction is not None:
                topicFunction(payload)
        
        # shutdown attached node process      
        @self.sio.event
        def shutdown():
            print("shutting down node")
            os._exit(0)
                
            
        def tryconnect():
            try:
                self.sio.connect(url)
            except Exception as e:
                print(f"SocketIO error: {e}")
        
        if not skip_connection:       
            tryconnect()

        # Install a global exception hook so any unhandled exception in this
        # process is reported to the server before the process exits.
        _node_ref = self

        def _global_exception_handler(exc_type, exc_value, exc_tb):
            """Report fatal errors to the behaviour tree via the server."""
            error_msg = "".join(traceback.format_exception(exc_type, exc_value, exc_tb))
            print(f"[Node:{node_name}] FATAL ERROR:\n{error_msg}")
            try:
                if _node_ref.connected:
                    _node_ref.sio.emit("node_error", {
                        "node":  node_name,
                        "error": str(exc_value) or repr(exc_value),
                    })
                    # Give the emit a moment to flush before the process dies
                    import time
                    time.sleep(0.5)
            except Exception:
                pass  # Don't recurse into another error
            # Call the default handler (prints traceback)
            sys.__excepthook__(exc_type, exc_value, exc_tb)

        sys.excepthook = _global_exception_handler
    
    # send data on topic
    def send(self, topic, data=None):
        if self.connected:
            print(f"Sending data ({data}) to topic: ({topic})")
            self.sio.emit("send_data", [topic, data])
        else:
            print("Node not connected!")
    
    # subscribe to topic
    def subscribe(self, topic, function):
        if self.connected:
            self.sio.emit("join_topic", topic)
            self.topics[topic] = function
        else:
            print("Node not connected!")
    
    # emit socket event    
    def emit(self, event):
        if self.connected:
            self.sio.emit(event)
        else:
            print("Node not connected!")

    # report a handled error to the server (behaviour tree will halt)
    def report_error(self, message: str):
        """
        Call this from any node to report a non-fatal but significant error
        that should halt the behaviour tree (e.g. sensor failure, hardware
        fault) without crashing the process.
        """
        print(f"[Node:{self.node_name}] Reporting error: {message}")
        if self.connected:
            self.sio.emit("node_error", {
                "node":  self.node_name,
                "error": message,
            })

File: Camera/Node.cpp
// Node.cpp
#include "Node.h"

SocketNode::SocketNode(const std::string& node_name, const std::string& url) : node_name(node_name) {
    // Bind events
    sio_client.set_open_listener(std::bind(&SocketNode::on_connect, this));
    sio_client.set_fail_listener([]() {
        std::cerr << "Socket.IO connection failed" << std::endl;
    });

    // Listen for "get_data" event
    sio_client.socket()->on("get_data", [this](sio::event& ev) {
        // Extract topic and payload from ev.get_message()
        if (ev.get_message()->get_flag() != sio::message::flag_array || ev.get_message()->get_vector().size() < 2) {
            return;
        }

        std::string topic = ev.get_message()->get_vector()[0]->get_string();
        std::string payload = ev.get_message()->get_vector()[1]->get_string();

        std::cout << "Received data on topic '" << topic << "': " << payload << std::endl;

        auto it = topics.find(topic);
        if (it != topics.end()) {
            it->second(payload);
        }
    });
    
    // Listen for shutdown
    sio_client.socket()->on("shutdown", [this](sio::event& ev) {
        this->on_shutdown();
    });

    try_connect(url);
}

SocketNode::~SocketNode() {
    sio_client.sync_close();
    sio_client.clear_con_listeners();
}

void SocketNode::on_connect() {
    std::cout << "Node " << node_name << " connected" << std::endl;
    connected = true;
    sio_client.socket()->emit("connect_node", node_name);
}

void SocketNode::on_get_data(const std::string& name, sio::message::ptr const& data,
                       bool need_ack, sio::message::ptr& ack_message) {
    if (data->get_flag() != sio::message::flag::flag_array || data->get_vector().size() < 2) {
        return;
    }

    std::string topic = data->get_vector()[0]->get_string();
    std::string payload = data->get_vector()[1]->get_string();

    std::cout << "Received data on topic '" << topic << "': " << payload << std::endl;

    auto it = topics.find(topic);
    if (it != topics.end()) {
        it->second(payload);
    }
}

void SocketNode::on_shutdown() {
    std::cout << "Shutting down node" << std::endl;
    std::exit(0);
}

void SocketNode::try_connect(const std::string& url) {
    try {
        sio_client.connect(url);
    } catch (const std::exception& e) {
        std::cerr << "Socket.IO connection error: " << e.what() << std::endl;
    }
}

void SocketNode::send(const std::string& topic, const std::string& data) {
    if (!connected) {
        std::cerr << "Node not connected!" << std::endl;
        return;
    }

    std::cout << "Sending data (" << data << ") to topic: (" << topic << ")" << std::endl;

    sio::message::ptr msg_array = sio::array_message::create();
    
    // Use get_vector() — this works in almost all versions
    auto& vec = msg_array->get_vector();
    vec.push_back(sio::string_message::create(topic));
    vec.push_back(sio::string_message::create(data));

    sio_client.socket()->emit("send_data", msg_array);
}

void SocketNode::subscribe(const std::string& topic, std::function<void(const std::string&)> function) {
    if (!connected) {
        std::cerr << "Node not connected!" << std::endl;
        return;
    }

    sio_client.socket()->emit("join_topic", topic);
    topics[topic] = function;
}

void SocketNode::emit(const std::string& event) {
    if (!connected) {
        std::cerr << "Node not connected!" << std::endl;
        return;
    }
    sio_client.socket()->emit(event);
}
